### ISSUE TITLE
Multiple Workflows vs. 1 main workflow

### ISSUE THREAD

**Issue by jluethi** [fractal-analytics-platform/fractal-server#236]
In writing the web GUI requirements and thinking more about the workflow flexibility priority, one topic that came up for me is the question of how we're handling flexibility of workflows [see multi-part workflows (e.g. running one part first, a second part later, see here for a bit: https://github.com/fractal-analytics-platform/fractal-server/issues/14), rerunning workflows (https://github.com/fractal-analytics-platform/fractal-server/issues/103) & running workflows on data subsets (https://github.com/fractal-analytics-platform/fractal-server/issues/109)] vs. workflow history (https://github.com/fractal-analytics-platform/fractal-server/issues/108) & easy import/export (https://github.com/fractal-analytics-platform/fractal-server/issues/1). The overarching question for me: Does a project have one main workflow or have multiple workflows?

On how we run things, I think there will often be different workflows, e.g. a user built a workflow for preprocessing, later wants to submit a workflow for segmentation & measurements. Likely this second part will be submitted as its own workflow that just contains the necessary tasks.
Or when a user wants to experiment with parameters and run a specific task just on a subset of the data (see https://github.com/fractal-analytics-platform/fractal-server/issues/109): We'd likely submit a workflow that just contains this one task and a different dataset (e.g. a dataset just limited to the well the user wants to process?).

This quickly leads to a project having dozens of workflows and it becomes hard to keep the overview, even harder to design a good interface for it. One thing I started thinking about and would like to discuss here:
Can projects have 1 main workflow?

Everything that is run on the whole input dataset becomes part of the workflow. If a user wants to run only part of a workflow, the new subset is generated on-the-fly from the main workflow, but the project still just has its one main workflow. If a user wants to run on a subset of the data, the workflow is submitted with a different input & output dataset on-the-fly, but the main workflow maintains its input & output data.

If this is feasible, it will keep it much cleaner for the user, we don't need to manage many workflows in the interface (even if we actually create them on-the-fly to submit subsets of workflows etc.) and workflow export & sharing will become much simpler

**Comment by jacopo-exact** [comment]
> Does a project have one main workflow or have multiple workflows?

projects have multiple workflows. As of right now, there is no way to elect one of them to "main workflow". This could be achieved by

- [ ] adding `Project.main_workflow_id: Optional[int]`

> On how we run things, I think there will often be different workflows, e.g. a user built a workflow for preprocessing, later wants to submit a workflow for segmentation & measurements. Likely this second part will be submitted as its own workflow that just contains the necessary tasks. Or when a user wants to experiment with parameters and run a specific task just on a subset of the data (see #109): We'd likely submit a workflow that just contains this one task and a different dataset (e.g. a dataset just limited to the well the user wants to process?).

the latter seems like the answer to me.
please note that this use case was the reason why we built version 0 with the complex nested structure of tasks, in which each task could be run independently as if it were a workflow on its own. we decided just a few weeks ago that we could let go this complexity in favour of a leaner flat structure, at the cost that the user must create a workflow even just for a single task.

> This quickly leads to a project having dozens of workflows and it becomes hard to keep the overview, even harder to design a good interface for it. One thing I started thinking about and would like to discuss here: Can projects have 1 main workflow?
> 
> Everything that is run on the whole input dataset becomes part of the workflow. If a user wants to run only part of a workflow, the new subset is generated on-the-fly from the main workflow, but the project still just has its one main workflow. If a user wants to run on a subset of the data, the workflow is submitted with a different input & output dataset on-the-fly, but the main workflow maintains its input & output data.

Workflows have no idea of what dataset(s) they will run on. A workflow is just a list of tasks that will be executed sequentially on whatever input is passed. This is by construction, as it would make portability from project to project much harder otherwise.

> If this is feasible, it will keep it much cleaner for the user, we don't need to manage many workflows in the interface (even if we actually create them on-the-fly to submit subsets of workflows etc.) and workflow export & sharing will become much simpler

We can certainly automate part of this. However, workflows still need to be persisted in the DB and associated explicitly with a project (this was one of the an explicit requirement that sparked the v1.x refactoring).

**Comment by jluethi** [comment]
Thanks for the feedback @jacopo-exact !

In that case, I think we should be working towards a "main workflow" for each project. 

> at the cost that the user must create a workflow even just for a single task.

That's fine. We could trigger the creation of such "subset-workflows" whenever a user chooses to submit only part of the workflow. There would still be just one main workflow, but when a user makes the decisions to run only tasks 1-3 out of 7 of the workflow, a new workflow with that subset is created and submitted. The user never needs to actively interact with this new subset workflow though.

> Workflows have no idea of what dataset(s) they will run on.

Great, that simplifies all the questions of running on subsets of the dataset then. It's a separate concern of deciding which datasets exist.

> We can certainly automate part of this. However, workflows still need to be persisted in the DB and associated explicitly with a project (this was one of the an explicit requirement that sparked the v1.x refactoring).

Sounds like a good plan to me. We can be explicit with this for testing and then figure out good ways, mostly for the web interface, to create & submit subsets of workflows.

**Comment by jluethi** [comment]
Thinking of this in the context of multiplexing workflows (https://github.com/fractal-analytics-platform/fractal-tasks-core/issues/199) and some of the complexity of input inference (https://github.com/fractal-analytics-platform/fractal-tasks-core/issues/200) and rerunning workflows:

We may want to have specific sections of the workflows, e.g.:
1. Preprocessing (image to image transformations)
2. Registration (only for multiplexing experiments)
3. Image Analysis (image to labels/tables)

It could make sense to have separate workflows for each that are chained after one another (to be discussed). Would be interesting if we can still represent it to the user as their "1 main workflow", especially in the web interface. 

tldr: I still like a 1 main workflow idea, but open to discuss how much we stick to it

**Comment by jluethi** [comment]
We have a good approach for this in Fractal V2 now
